rs-matter/src/data_model/subscriptions.rs

```diff
@@ -0,0 +1,142 @@
+/*
+ *
+ *    Copyright (c) 2020-2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+use core::cell::RefCell;
+
+use embassy_sync::blocking_mutex::raw::NoopRawMutex;
+use embassy_time::Instant;
+
+use portable_atomic::{AtomicU32, Ordering};
+
+use crate::utils::notification::Notification;
+
+struct Subscription {
+    node_id: u64,
+    id: u32,
+    min_int_secs: u16,
+    max_int_secs: u16,
+    reported_at: Instant,
+    changed: bool,
+}
+
+impl Subscription {
+    pub fn report_due(&self, now: Instant) -> bool {
+        self.changed
+            && self.reported_at + embassy_time::Duration::from_secs(self.min_int_secs as _) <= now
+    }
+
+    pub fn is_expired(&self, now: Instant) -> bool {
+        self.reported_at + embassy_time::Duration::from_secs(self.max_int_secs as _) <= now
+    }
+}
+
+/// A utility for tracking subscriptions accepted by the data model.
+///
+/// The `N` type parameter specifies the maximum number of subscriptions that can be tracked at the same time.
+/// Additional subscriptions are rejected by the data model with a "respource exhausted" IM status message.
+pub struct Subscriptions<const N: usize> {
+    next_subscription_id: AtomicU32,
+    subscriptions: RefCell<heapless::Vec<Subscription, N>>,
+    pub(crate) notification: Notification<NoopRawMutex>,
+}
+
+impl<const N: usize> Subscriptions<N> {
+    /// Create the instance.
+    #[inline(always)]
+    pub const fn new() -> Self {
+        Self {
+            next_subscription_id: AtomicU32::new(1),
+            subscriptions: RefCell::new(heapless::Vec::new()),
+            notification: Notification::new(),
+        }
+    }
+
+    /// Notify the instance that some data in the data model has changed and that it should re-evaluate the subscriptions
+    /// and report on those that concern the changed data.
+    ///
+    /// This method is supposed to be called by the application code whenever it changes the data model.
+    pub fn notify_changed(&self) {
```

Comment:
should there be some TODO argument so we pass in "what changed"? That way we could conditionally update subscription dirty status on paths overlaps only.

Reply:
As mentioned elsewhere, I would rather imagine that upon being notified that "something had changed", the subscriptions' code itself will detect - for each subscription - whether that particular subscription is affected by the changes. One memory efficient way to do this is to read the `data_ver` of all clusters reported via that subscription, shuffle them in an md5 or sha1 hash, and report only if the md5 / sha1 hash is different from the "previous" one which is kept inside the subscription struct itself.